src/Microsoft.Tye.Hosting/ReplicaRegistry.cs

```diff
@@ -0,0 +1,126 @@
+            _store = new StateStore(application);
+        }
+
+        public async Task Reset()
+        {
+            await PurgeAll();
+            _store.Reset(true, true);
+        }
+
+        public async Task WriteReplicaEvent(ReplicaEvent replicaEvent)
+        {
+            var serviceType = replicaEvent.Replica.Service.ServiceType;
+            if (_replicaInstantiators.TryGetValue(serviceType, out var instantiator))
+            {
+                var serialized = await instantiator.SerializeReplica(replicaEvent);
+                await _store.WriteEvent(new StoreEvent
+                {
+                    ServiceType = serviceType,
+                    State = replicaEvent.State,
+                    SerializedEvent = serialized
+                });
+            }
+        }
+
+        private async Task PurgeAll()
+        {
+            var events = await _store.GetEvents();
+            var tasks = new List<Task>(events.Count);
+
+            foreach (var @event in events)
+            {
+                if (_replicaInstantiators.TryGetValue(@event.ServiceType, out var instantiator))
+                {
+                    var replica = await instantiator.DeserializeReplicaEvent(@event.SerializedEvent);
+                    tasks.Add(instantiator.HandleStaleReplica(replica));
+                }
+            }
+
+            await Task.WhenAll(tasks);
+        }
+
+        public void Dispose()
+        {
+            _store.Dispose();
+        }
+
+        private struct StoreEvent
+        {
+            public ServiceType ServiceType { get; set; }
+            public ReplicaState State { get; set; }
+            public IDictionary<string, string> SerializedEvent { get; set; }
+        }
+
+        private class StateStore : IDisposable
+        {
+            //TOOD: Consider Sqlite...
+
+            private string _tyeFolderPath;
+            private string _eventsFile;
+
+            public StateStore(Model.Application application)
+            {
+                _tyeFolderPath = Path.Join(Path.GetDirectoryName(application.Source), ".tye");
+                Reset(false, true);
+
+                _eventsFile = Path.Join(_tyeFolderPath, "events");
+            }
+
+            public async Task WriteEvent(StoreEvent @event)
+            {
+                var contents = JsonSerializer.Serialize(@event, options: new JsonSerializerOptions
+                {
+                    WriteIndented = false
+                });
+                await File.AppendAllTextAsync(_eventsFile, contents + Environment.NewLine);
+            }
+
+            public async ValueTask<IList<StoreEvent>> GetEvents()
+            {
+                if (!File.Exists(_eventsFile))
+                {
+                    return Array.Empty<StoreEvent>();
+                }
+
+                var contents = await File.ReadAllTextAsync(_eventsFile);
+                var events = contents.Split(Environment.NewLine);
```

Comment:
Are the events forbidden from containing a newline?

Reply:
I suppose that's the case if you are using `WriteIndented = false` since newlines in the strings in JSON would be escaped already.

Reply:
Yeah. The `WriteIndented = false` should prevent line breaks and other indentations.